src/SyGuS.ml

```diff
@@ -131,8 +124,19 @@ let parse_sexps (sexps : Sexp.t list) : t =
               | List [ (Atom "inv-constraint") ; (Atom _invf_name) ; (Atom _pref_name)
                                                ; (Atom _transf_name) ; (Atom _postf_name) ]
                 -> pref_name := _pref_name ; transf_name := _transf_name ; postf_name := _postf_name
-                 ; if not (String.equal !invf_name _invf_name)
-                   then raise (Parse_Exn ("Invariant function [" ^ _invf_name ^ "] not declared"))
+                 ; (if not (String.equal !invf_name _invf_name)
+                    then raise (Parse_Exn ("Invariant function [" ^ _invf_name ^ "] not declared")))
+                 ; (if not (String.equal !pref_name _pref_name)
+                    then raise (Parse_Exn ("Precondition function [" ^ _pref_name ^ "] not declared")))
+                 ; (if not (String.equal !transf_name _transf_name)
+                    then raise (Parse_Exn ("Transition function [" ^ _transf_name ^ "] not declared")))
+                 ; (if not (String.equal !postf_name _postf_name)
+                    then raise (Parse_Exn ("Postcondition function [" ^ _postf_name ^ "] not declared")))
+                 ; (match List.map ~f:(fun name -> func_from_funcs !funcs name) [_pref_name; _transf_name; _postf_name] with
+                   | [[pref];[transf];[postf]] -> variables := transf.args ;
+                                                  if not ((check_arg_order !variables pref) && (check_arg_order !variables postf))
+                                                  then raise (Parse_Exn ("Argument order not consistent between pre, trans, and post functions"))
+                   | _ -> raise (Parse_Exn ("Multiple definitions of a pre, trans, or post function.")) (* Should be impossible to get here. *))
```

Comment:
1) Primed variables (those with `!` suffix) do not appear in pre and post. So simply comparing `transf.args` to `.args` of other functions is incorrect!

2) Multiple definitions are already checked at `define-fun` case.
Moreover, these 5 lines can be simplified a lot.

Reply:
The 5 lines can be replaced with: (adjust indentation after copying)

```ocaml
; let variables := !invf_vars @ (List.map !invf_vars ~f:(fun (v,t) -> (v^"!",t))) in
  let [pref ; transf ; postf] =
    List.map [_pref_name ; _transf_name ; _postf_name]
             ~f:(fun name -> List.find !funcs ~f:(fun f -> String.equal f.name name))
   in (match pref with Some pref -> if not (String.equal pref.args !invf_vars)
                                    then raise (Parse_Exn "[" ^ _invf_name ^ "] and [" ^ _pref_name ^ "] have inconsistent arguments!")
                     | _ -> raise (Parse_Exn "Precondition [" ^ _pref_name ^ "] not defined!"))
    ; (match postf with Some postf -> if not (String.equal postf.args !invf_vars)
                                      then raise (Parse_Exn "[" ^ _invf_name ^ "] and [" ^ _postf_name ^ "] have inconsistent arguments!")
                      | _ -> raise (Parse_Exn "Postcondition [" ^ _postf_name ^ "] not defined!"))
    ; (match transf with Some transf -> if not (String.equal transf.args !variables)
                                        then raise (Parse_Exn "[" ^ _invf_name ^ "] and [" ^ _transf_name ^ "] have inconsistent arguments!")
                       | _ -> raise (Parse_Exn "Transition function [" ^ _transf_name ^ "] not defined!"))
```

I haven't tested this code, but I hope you get the idea. Lemme know if you don't understand.

Reply:
You also don't need the two small helper functions that you defined above. You should remove those.